Update to ESR 91 and add base patches for Sailfish OS

llewelld opened the pull request:
Updates the submodule mirror to ESR 91. Adds patches for building for Sailfish OS.

There are some things to note:
1. I've only tested with the aarch64 build. The arm7hl and i486 builds will likely need some tweaks.
2. The spec file has a couple of dependencies that rely on the version of the build engine I have installed, which means GCC 8.3.0 and libclang 10. These almost certainly need updating for the latest build engine (but this should be a simple change to the spec file).
3. This is only intended as a base. There are a bunch of refinements that'll inevitably be needed.
4. This also relies on updates to all the other pieces: sailfish-browser, qtmozembed, embedlite-components, sailfissh-components-webview. I'll post links to the relevant PRs for these once they're available.

llewelld commented:
The full set of pull requests that go together:

1. https://github.com/sailfishos/gecko-dev/pull/162
2. https://github.com/sailfishos/qtmozembed/pull/49
3. https://github.com/sailfishos/sailfish-browser/pull/1015
4. https://github.com/sailfishos/embedlite-components/pull/100
5. https://github.com/sailfishos/sailfish-components-webview/pull/169

rainemak commented on rpm/0005-sailfishos-gecko-Fix-build-version-requirements.patch:
```diff
@@ -0,0 +1,57 @@
+     if js_standalone:
+         return "nspr >= 4.10"
+-    return "nspr >= 4.32"
++    return "nspr >= 4.29"
```
This looks fishy as we already have 4.35 => looks like a something old.

mlehtima replied:
That was still needed in 4.5.0 so maybe it's because of that.

llewelld replied:
Thanks for noticing @rainemak and for clarifying the reason @mlehtima; you[re both right.

After reverting this change and building for the 4.6 target there weren't any issues, so it's now back to the original 4.3.2.

rainemak commented on rpm/xulrunner-qt5.spec:
```diff
@@ -57,122 +57,99 @@ URL:        https://git.sailfishos.org/mer-core/gecko-dev
 BuildRequires:  cargo
 BuildRequires:  pkgconfig(Qt5Quick)
 BuildRequires:  pkgconfig(Qt5Network)
+BuildRequires:  pkgconfig(Qt5Widgets)
```
Why do we need this?

llewelld replied:
I'm not sure what the reason was, but the packages build fine without this so I've removed this line.

mlehtima replied:
There seems to be a requirement for that in gfx/cairo/cairo/src/cairo-qt-surface.cpp

llewelld replied:
Thanks @mlehtima; I've removed the revert commit.

rainemak reviewed (commented):
Looks good. Few minor remarks added.

rainemak commented on rpm/0045-sailfishos-gecko-Add-and-adjust-embedlite-static-pre.patch:
```diff
@@ -0,0 +1,152 @@
++# Make gecko compositor use GL context/surface provided by the application
++-   name: embedlite.compositor.external_gl_context
++    type: bool
++    value: true
```
I see browser as a special case. Thus, default should be false and browser set this to true.

https://github.com/sailfishos/sailfish-components-webview/pull/169/commits/234f548ff61f721cb06b4c28ebe86235870d87d9#r1698347125

llewelld replied:
Okay; I switched this to `false` and it's now set in the browser initialisation code. Is this what you were thinking?

rainemak replied:
Yeap, defaulting to false was exactly what I was thinking about.

rainemak replied:
It's perfectly like this as well so that sailfish-components-webview and sailfish-browser do alter prefs. Let's create more PRs on top as needed.

llewelld commented on embedding/embedlite/config/mozconfig.merqtxulrunner:
```diff
@@ -24,7 +25,8 @@ ac_add_options --with-gl-provider=EGL
 ac_add_options --enable-update-channel=nightly
 ac_add_options --disable-tests
-ac_add_options --disable-marionette
+#ac_add_options --disable-marionette
```
Yes, this flag was combined into the `--disable-webrender` flag, see the following:
1. https://firefox-source-docs.mozilla.org/testing/marionette/Building.html (final paragraph)
2. https://phabricator.services.mozilla.com/D115583

Thanks for flagging up the commented code; I've removed the line completely now.

llewelld commented on embedding/embedlite/embedding.js:
```diff
@@ -108,23 +108,12 @@ pref("media.gstreamer.enable-blacklist", false);
-
 // Make gecko compositor use GL context/surface provided by the application.
-pref("embedlite.compositor.external_gl_context", false);
+//pref("embedlite.compositor.external_gl_context", false);
```
Thanks; I've removed them.

llewelld commented on embedding/embedlite/embedshared/EmbedLiteViewChild.cpp:
```diff
@@ -517,18 +479,22 @@ EmbedLiteViewChild::WebWidget()
 /*----------------------------TabChildIface-----------------------------------------------------*/
 
-mozilla::ipc::IPCResult EmbedLiteViewChild::RecvLoadURL(const nsString &url)
+mozilla::ipc::IPCResult EmbedLiteViewChild::RecvLoadURL(const nsString &url, const bool& aFromExternal)
```
Done.

llewelld commented on embedding/embedlite/embedshared/EmbedLiteViewChild.cpp:
```diff
@@ -614,16 +580,24 @@ mozilla::ipc::IPCResult EmbedLiteViewChild::RecvSetIsActive(const bool &aIsActiv
+              : dom::ExplicitActiveStatus::Inactive
+  );
+
+  //presShell->ActivenessMaybeChanged();
```
You're right, thanks for noticing. I've removed it.

llewelld commented on embedding/embedlite/embedthread/EmbedContentController.cpp:
```diff
@@ -254,7 +254,8 @@ void EmbedContentController::NotifyFlushComplete()
 
-void EmbedContentController::NotifyPinchGesture(PinchGestureInput::PinchGestureType aType, const EmbedContentController::ScrollableLayerGuid &aGuid, LayoutDeviceCoord aSpanChange, Modifiers aModifiers)
+
+void EmbedContentController::NotifyPinchGesture(PinchGestureInput::PinchGestureType aType, const EmbedContentController::ScrollableLayerGuid& aGuid, const LayoutDevicePoint& aFocusPoint, LayoutDeviceCoord aSpanChange, Modifiers aModifiers)
```
Done.

llewelld commented on embedding/embedlite/embedthread/EmbedContentController.h:
```diff
@@ -51,8 +51,10 @@ class EmbedContentController : public mozilla::layers::GeckoContentController
                                     int aArg = 0) override;
   virtual void NotifyFlushComplete() override;
 
+
```
Thanks; removed.

llewelld commented on embedding/embedlite/embedthread/EmbedLiteCompositorBridgeParent.cpp:
```diff
@@ -16,20 +16,24 @@
 #include "VsyncSource.h"
 
+
+
```
Removed.

llewelld commented on rpm/0045-sailfishos-gecko-Add-and-adjust-embedlite-static-pre.patch:
```diff
@@ -0,0 +1,152 @@
++# this pref only makes sense when using an external compositor gl context
++-   name: embedlite.compositor.request_external_gl_context_early
++    type: bool
++    value: true
```
I've also changed this. Please say if I misunderstood what you were after.

rainemak replied:
You didn't misunderstood. However, I'm still seeing this as true.

mlehtima commented:
Some comments about content of the patches, the gecko version change in "Fix specfile versions" should be moved to the initial commit "Update spec for ESR 91.9.0" so it's not included in a hack commit. Also the commits "Fix 32-bit builds" and "Disable patch "Fix 32-bit builds" should be removed and the patch included only along with the other patches in the last commit "Add patches for ESR 91". Another thing is rebasing the changes on top of master branch of sailfishos/gecko-dev which contains just some changes for newer clang, gcc, glibc and rust. The rebase is actually quite simple and of the new patches in master branch only 4 remain after rebasing those on top of the current esr91 branch (patches 101, 102, 109 and 110)

llewelld commented:
> Some comments about content of the patches...

Thanks for checking this so carefully @mlehtima. I've made all of the changes:
1. Squashed "Fix specfile versions" into "Update spec for ESR 91.9.0".
2. Removed the "Fix 32-bit builds" commits with the patch added allongside all the others.
3. Rebased.

For 1, the following change is included in the commit, but this change may only be necessary because I'm using an older build engine (as opposed to an older target). So perhaps this should be removed (or updated):
```patch
-SBOX_DISABLE_MAPPING=1 cp /usr/lib/libclang.so.13 "%BUILD_DIR"/lib/
+SBOX_DISABLE_MAPPING=1 cp /usr/lib/libclang.so.10 "%BUILD_DIR"/lib/
```

For 2, I rebased on top of `sailfishos-esr9`, which I think is what you meant. I tried rebasing on top of `master` but it caused a bit of a mess! :grin:

llewelld commented:
Unfortunately the changes I made to rearrange the preference managed to break things. I've reverted them so that `embedlite.compositor.external_gl_context` and `embedlite.compositor.request_external_gl_context_early` now default to `true`. This means that I've removed the assignment from sailfish-browser (see https://github.com/sailfishos/sailfish-browser/pull/1074/commits/53d3f0ad98158e4bba38e87950eddbc827996743) and have restored the code that amends them in sailfish-components-webview (see https://github.com/sailfishos/sailfish-components-webview/pull/169/commits/4c2429e24deaf136ed35a955bfd95248b204379f). It could be there's a better way to make this work and I just arranged things incorrectly but I'll need to look into it a bit further to check.

llewelld commented:
Speculatively it looks to me like it might be a timing issue. Setting `embedlite.compositor.external_gl_context` and `embedlite.compositor.request_external_gl_context_early` in `DeclarativeWebUtils::updateWebEngineSettings()` (as I was doing) for the browser is too late in the startup sequence. The renderer starts up before this points triggering the crash.
